File: solutions/haskell/033-first-missing-positive.hs
{-
  Problem 33: First Missing Positive (LeetCode 41)
  Difficulty: Hard
  Language: Haskell
-}
import qualified Data.Set as Set

firstMissingPositive :: [Int] -> Int
firstMissingPositive nums =
  let s = Set.fromList (filter (> 0) nums)
      go i = if Set.member i s then go (i + 1) else i
  in go 1

File: solutions/ocaml/033-first-missing-positive.ml
(*
  Problem 33: First Missing Positive (LeetCode 41)
  Difficulty: Hard
  Language: OCaml
*)
let first_missing_positive nums =
  let module S = Set.Make(Int) in
  let s = List.fold_left (fun acc n ->
    if n > 0 then S.add n acc else acc) S.empty nums in
  let rec go i = if S.mem i s then go (i + 1) else i in
  go 1
